src/core/shards/events.cpp

```diff
@@ -0,0 +1,113 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2023 Fragcolor Pte. Ltd. */
+
+#include "shared.hpp"
+
+namespace shards {
+namespace Events {
+struct Base {
+  std::optional<std::reference_wrapper<entt::dispatcher>> _dispatcher;
+  std::string _eventName{"MyEvent"};
+
+  static SHTypesInfo inputTypes() { return CoreInfo::AnyType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::AnyType; }
+
+  static inline Parameters params{{"Name", SHCCSTR("The name of the event dispatcher to use."), {CoreInfo::StringType}}};
+
+  static SHParametersInfo parameters() { return params; }
+
+  void setParam(int index, const SHVar &value) {
+    switch (index) {
+    case 0:
+      _eventName = value.payload.stringValue;
+      break;
+    default:
+      break;
+    }
+  }
+
+  SHVar getParam(int index) {
+    switch (index) {
+    case 0:
+      return Var(_eventName);
+    default:
+      return Var::Empty;
+    }
+  }
+
+  SHTypeInfo compose(const SHInstanceData &data) {
+    _dispatcher = GetGlobals().Dispatchers[_eventName];
```

Comment:
I don't think you can keep this as reference, as soon as the collection decides to move stuff around (after adding/removing), the references will become invalid.
Might have to store the values inside a shared/unique_ptr in the map if you want to reference them by address

Reply:
it's a `unordered_map` the memory will not be compacted or moved by spec.

Reply:
Ah, never though about that, but it seems that way yes.

Reply:
It's a very very good clever trick you should use, and we use a lot even in channels.
`std::deque` also has this property which is why I use it sometimes vs vector (which invalidates)